src/libmongoc/src/mongoc/mcd-integer.h

```diff
@@ -0,0 +1,411 @@
+           then: min_fac * right = (MIN / right) * right
+           then: min_fac * right =  MIN
+         */
+         const int64_t min_fac = INT64_MIN / right;
+         if (left < min_fac) {
+            /**
+            Given: left         < min_fac
+             then: left * right < min_fac * right
+             with: MIN          = min_fac * right
+             then: left * right < MIN
+            */
+            return true;
+         } else {
+            /**
+            Given: left         >= min_fac
+             then: left * right >= min_fac * right
+             with: MIN          =  min_fac * right
+             then: left * right >= MIN
+            */
+            return false;
+         }
+      } else {
+         /**
+         Given: left  <= 0
+           and: right <= 0
+          then: left * right <= 0
+          THEN: left * right <  MAX
+         */
+         if (right == 0) {
+            // Multiplication by zero will never overflow
+            return false;
+         } else {
+            /**
+            Given: right <= 0
+              and: right != 0
+             then: right < 0
+            */
+            if (right == -1) {
+               /**
+               Given: right        = -1
+                 and: N * -1       = -N
+                then: left * right = -left
+               */
+               if (left == INT64_MIN) {
+                  /**
+                  Given: left  = MIN
+                    and: -MIN  = MAX + 1  [By definition of 2's complement]
+                   then: -left = MAX + 1
+                   then: left * right = MAX + 1
+                   THEN: left * right > MAX
+                  */
+                  return true;
+               } else {
+                  /**
+                  Given: left >= MIN
+                    and: left != MIN
+                   then: left >  MIN
+
+                  Given: -MIN  =  MAX + 1 [By definition of 2's complement]
+                    and: left  >  MIN
+                   then: -left <  MAX + 1
+                   THEN: -left <= MAX
+                   */
+                  return false;
+               }
+            } else {
+               /**
+               Given: right <   0
+                 and: right != -1
+                then: right <  -1
+
+               Define: min_fac         =  MIN / right
+                 then: min_fac * right = (MIN / right) * right
+                 then: min_fac * right =  MIN
+               */
+               const int64_t min_fac = INT64_MIN / right;
+               if (left < min_fac) {
+                  /**
+                  Given: left         < min_fac
+                    and:        right < 0
+                   then: left * right > min_fac * right
+                    b/c: [Mult by negative flips ordering]
+                   with: MIN          = min_fac * right
+                   THEN: left * right > MIN
+                  */
+                  return false;
```

Comment:
Strongly recommend adding a comprehensive test suite (e.g. `test-mcd-integer.c`) to ensure overflow checks behave as expected given their complexity. e.g. `_mcd_i64_mul_would_overflow (-1, INT64_MIN) -> false` is incorrect.

Reply:
Fool me twice,